Rover/sailboat.cpp

```diff
@@ -260,6 +264,57 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
         wingsail_out *= -1.0f;
     }
 
+    //
+    // direct mast rotation control
+    // configure SERVOx_MAX/MIN such that the mast rotates +-90 degrees
+    //
+
+    if (!is_positive(desired_speed)) {
+        // rotating sails can be used to reverse, but not in this version
+        mast_rotation_out = 0.0f;
+    } else {
+        // + is wind over starboard side, - is wind over port side
+        float wind_dir_apparent_signed = degrees(rover.g2.windvane.get_apparent_wind_direction_rad());
+        float wind_dir_apparent_abs = fabsf(wind_dir_apparent_signed);
+        float wind_dir_apparent_sign = wind_dir_apparent_signed / wind_dir_apparent_abs;
+
+        // compute absolute sail rotation
+        float mast_rotation_angle = 0.0f;
+        bool mast_rotation_necessary = true;
+        if (wind_dir_apparent_abs < sail_angle_ideal) {
+            // in irons, center the sail.
+            mast_rotation_angle = 0.0f;
+        } else if (wind_dir_apparent_abs < (90.0f + sail_angle_ideal)) {
+            // use sail as a lift device, at ideal angle of attack
+            mast_rotation_angle = (wind_dir_apparent_abs - sail_angle_ideal);
+        } else {
+            // use sail as drag device
+            mast_rotation_angle = 90.0f;
+            // but avoid wagging the sail as the wind oscillates
+            // between 180 and -180 degrees
+            if (wind_dir_apparent_abs > 135.0f &&
+                is_equal(fabsf(SRV_Channels::get_output_norm(SRV_Channel::k_mast_rotation)), 1.0f)) {
+                // sail is already at either 90 or -90 degrees, leave
+                // it where it is; drag is identical both ways
+                mast_rotation_necessary = false;
```

Comment:
```suggestion
mast_rotation_out = SRV_Channels::get_output_scaled(SRV_Channel::k_mast_rotation) * sail_angle_max;

```

You don't need the flag then

Reply:
You will need to set the sign/tack to match too

Reply:
I'm not sure I understand how this would work. You didn't like the NaN, so I replaced it with a flag. Are you trying to get rid of the flag now? You can do that, but it would take a lot more restructuring of the logic. If I read your suggestion correctly, you want to set `mast_rotation_out`, which is in percent, in the middle of angular calculations. In addition to confusing the structure, this assignment would get overwritten in the next block, where I convert from degrees to percent. You'd need a flag to prevent that from happening, which is what I have now without mixing degrees and percent.

Your second remark is more obscure. When running, the sign is determined by hysteresis, not tack. This is the whole purpose of the special case with the `rotation_necessary` flag--to ignore the tack in this one case.

The structure of my code is like this:

* Compute optimum absolute _angle_ of attack depending on the point of sail. The calculation is the same for port and starboard tacks in all cases except dead downwind. Everything is in degrees.
* Put the sign back in. Everything is still in degrees.
* Scale from degrees to percent.
* For the special case for reducing hysteresis when running downwind, keep current output. This part is entirely in percent.

If you want me to restructure this, please let me know.

Reply:
Yeah, I think you can remove the flag and simplify the code.

Reply:
If you want to propose a different structure for the code, I am happy to incorporate it into my PR.